HW2/P3/mandelbrot.pyx

```diff
@@ -32,17 +70,50 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
 
     with nogil:
-        for i in range(in_coords.shape[0]):
-            for j in range(in_coords.shape[1]):
-                c = in_coords[i, j]
-                z = 0
+        for i in prange(in_coords.shape[0], num_threads=NUM_THREADS, schedule='static', chunksize=1):
+            for j in range(0, in_coords.shape[1], 8):
+                iter_count = AVX.float_to_float8(0.)
+                # convert data into float8
+                c_real = AVX.make_float8(in_coords[i, j].real,
+                                         in_coords[i, j+1].real,
+                                         in_coords[i, j+2].real,
+                                         in_coords[i, j+3].real,
+                                         in_coords[i, j+4].real,
+                                         in_coords[i, j+5].real,
+                                         in_coords[i, j+6].real,
+                                         in_coords[i, j+7].real)
+                c_imag = AVX.make_float8(in_coords[i, j].imag,
+                                         in_coords[i, j+1].imag,
+                                         in_coords[i, j+2].imag,
+                                         in_coords[i, j+3].imag,
+                                         in_coords[i, j+4].imag,
+                                         in_coords[i, j+5].imag,
+                                         in_coords[i, j+6].imag,
+                                         in_coords[i, j+7].imag)
+                z_real = AVX.float_to_float8(0.)
+                z_imag = AVX.float_to_float8(0.)
                 for iter in range(max_iterations):
-                    if magnitude_squared(z) > 4:
+                    # compute abs of float8
+                    z_real_sq = AVX.mul(z_real, z_real)
+                    z_imag_sq = AVX.mul(z_imag, z_imag)
+                    magnitudes_squared = AVX.add(z_real_sq, z_imag_sq)
+                    # check if abs of float8 is less greater than 4
+                    mask = AVX.less_than(magnitudes_squared, AVX.float_to_float8(4.))
+                    # if float8 all greater than 4
+                    if not AVX.signs(mask): 
                         break
-                    z = z * z + c
-                out_counts[i, j] = iter
-
-
+                    # increment counts where magnitude_squared is less than 4
+                    iter_count = AVX.add(iter_count, AVX.bitwise_and(mask, AVX.float_to_float8(1.)))
+                    # z = z * z + c
+                    # z_real = z_real**2 - z_imag**2 + c_real
+                    # z_imag = 2 * z_real * z_imag * i + c_imag * i
+                    tmp_real = AVX.add(AVX.sub(z_real_sq, z_imag_sq), c_real)
+                    tmp_imag = AVX.add(AVX.mul(AVX.float_to_float8(2.), AVX.mul(z_real, z_imag)), c_imag)
+                    z_real, z_imag = tmp_real, tmp_imag
+                
+                AVX.to_mem(iter_count, &(out_vals[0]))
+                for k in range(8):
```

Comment:
I did this before too. But I realized if I store them in backwards for c_real,c_imag to start with, then I could just call AVX.to_mem(iter_count, &(out_vals[0])) without any for loop later. This could optimize the code a bit.